File: src/main/java/io/fabric8/launcher/booster/catalog/rhoar/RhoarBoosterCatalogService.java
package io.fabric8.launcher.booster.catalog.rhoar;

import java.io.BufferedReader;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.TreeSet;
import java.util.concurrent.ExecutorService;
import java.util.function.Predicate;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import javax.annotation.Nullable;

import io.fabric8.launcher.booster.catalog.AbstractBoosterCatalogService;
import io.fabric8.launcher.booster.catalog.BoosterDataTransformer;
import io.fabric8.launcher.booster.catalog.BoosterFetcher;
import io.fabric8.launcher.booster.catalog.YamlConstructor;
import io.fabric8.launcher.booster.catalog.spi.BoosterCatalogListener;
import io.fabric8.launcher.booster.catalog.spi.BoosterCatalogPathProvider;
import org.yaml.snakeyaml.Yaml;
import org.yaml.snakeyaml.representer.Representer;

public class RhoarBoosterCatalogService extends AbstractBoosterCatalogService<RhoarBooster> implements RhoarBoosterCatalog {

    private static final String METADATA_FILE = "metadata.yaml";

    private static final Logger logger = Logger.getLogger(RhoarBoosterCatalogService.class.getName());

    protected RhoarBoosterCatalogService(Builder config) {
        super(config);
    }

    @Override
    protected RhoarBooster newBooster(@Nullable Map<String, Object> data, BoosterFetcher boosterFetcher) {
        return new RhoarBooster(data, boosterFetcher);
    }

    @Override
    public Set<Mission> getMissions() {
        return toMissions(getPrefilteredBoosters());
    }

    @Override
    public Set<Mission> getMissions(Predicate<RhoarBooster> filter) {
        return toMissions(getPrefilteredBoosters().filter(filter));
    }

    @Override
    public Set<Runtime> getRuntimes() {
        return toRuntimes(getPrefilteredBoosters());
    }

    @Override
    public Set<Runtime> getRuntimes(Predicate<RhoarBooster> filter) {
        return toRuntimes(getPrefilteredBoosters().filter(filter));
    }

    @Override
    public Set<Version> getVersions(Predicate<RhoarBooster> filter) {
        return toVersions(getPrefilteredBoosters().filter(filter));
    }

    @Override
    public Set<Version> getVersions(Mission mission, Runtime runtime) {
        return getVersions(BoosterPredicates.withMission(mission).and(BoosterPredicates.withRuntime(runtime)));
    }

    @Override
    public Optional<RhoarBooster> getBooster(Mission mission, Runtime runtime, Version version) {
        return getPrefilteredBoosters()
                .filter(BoosterPredicates.withMission(mission))
                .filter(BoosterPredicates.withRuntime(runtime))
                .filter(BoosterPredicates.withVersion(version))
                .findAny();
    }

    private Set<Runtime> toRuntimes(Stream<RhoarBooster> bs) {
        return bs
                .filter(b -> b.getRuntime() != null)
                .map(RhoarBooster::getRuntime)
                .collect(Collectors.toCollection(TreeSet::new));
    }

    private Set<Mission> toMissions(Stream<RhoarBooster> bs) {
        return bs
                .filter(b -> b.getMission() != null)
                .map(RhoarBooster::getMission)
                .collect(Collectors.toCollection(TreeSet::new));
    }

    private Set<Version> toVersions(Stream<RhoarBooster> bs) {
        return bs
                .filter(b -> b.getVersion() != null)
                .map(RhoarBooster::getVersion)
                .collect(Collectors.toCollection(TreeSet::new));
    }

    @Override
    protected void indexBoosters(final Path catalogPath, final Set<RhoarBooster> boosters) throws IOException {
        super.indexBoosters(catalogPath, boosters);

        // Update the boosters with the proper info for missions, runtimes and versions
        Map<String, Mission> missions = new HashMap<>();
        Map<String, Runtime> runtimes = new HashMap<>();

        // Read the metadata for missions and runtimes
        Path metadataFile = catalogPath.resolve(METADATA_FILE);
        if (Files.exists(metadataFile)) {
            processMetadata(metadataFile, missions, runtimes);
        }

        for (RhoarBooster booster : boosters) {
            List<String> path = booster.getDescriptor().path;
            if (path.size() >= 3) {
                Runtime r = runtimes.get(path.get(0));
                if (r == null) {
                    r = new Runtime(path.get(0));
                    logger.log(Level.WARNING, "Runtime '{0}' not found in metadata", r.getId());
                }
                booster.setRuntime(r);

                Version v = r.getVersions().get(path.get(1));
                if (v == null) {
                    v = new Version(path.get(1));
                    logger.log(Level.WARNING, "Version '{0}' not found in Runtime '{1}' metadata", new Object[]{v.getId(), r.getId()});
                }
                booster.setVersion(v);

                Mission m = missions.get(path.get(2));
                if (m == null) {
                    m = new Mission(path.get(2));
                    logger.log(Level.WARNING, "Mission '{0}' not found in metadata", m.getId());
                }
                booster.setMission(m);
            }
        }
    }

    /**
     * Process the metadataFile and adds to the specified missions and runtimes
     * maps
     */
    @SuppressWarnings("unchecked")
    public void processMetadata(Path metadataFile, Map<String, Mission> missions, Map<String, Runtime> runtimes) {
        logger.info(() -> "Reading metadata at " + metadataFile + " ...");

        Representer rep = new Representer();
        rep.getPropertyUtils().setSkipMissingProperties(true);
        Yaml yaml = new Yaml(new YamlConstructor(), rep);
        try (BufferedReader reader = Files.newBufferedReader(metadataFile)) {
            Map<String, Object> metadata = yaml.loadAs(reader, Map.class);

            if (metadata.get("missions") instanceof List) {
                List<Map<String, Object>> ms = (List<Map<String, Object>>) metadata.get("missions");
                ms.stream()
                        .map(e -> new Mission(
                                (String) e.get("id"),
                                (String) e.get("name"),
                                (String) e.get("description"),
                                (Map<String, Object>) e.getOrDefault("metadata", Collections.emptyMap())))
                        .forEach(m -> missions.put(m.getId(), m));
            }

            if (metadata.get("runtimes") instanceof List) {
                List<Map<String, Object>> rs = (List<Map<String, Object>>) metadata.get("runtimes");
                rs.stream()
                        .map(e -> new Runtime(
                                (String) e.get("id"),
                                (String) e.get("name"),
                                (String) e.get("description"),
                                (Map<String, Object>) e.getOrDefault("metadata", Collections.emptyMap()),
                                (String) e.get("icon"),
                                getMetadataVersions(e.get("versions"))))
                        .forEach(r -> runtimes.put(r.getId(), r));
            }
        } catch (Exception e) {
            logger.log(Level.SEVERE, "Error while processing metadata " + metadataFile, e);
        }
    }

    @SuppressWarnings("unchecked")
    private Map<String, Version> getMetadataVersions(Object versionsList) {
        if (versionsList instanceof List) {
            Map<String, Version> versions = new HashMap<>();
            List<Map<String, Object>> vs = (List<Map<String, Object>>) versionsList;
            vs.stream()
                    .map(e -> new Version(
                            (String) e.get("id"),
                            (String) e.get("name"),
                            (String) e.get("description"),
                            (Map<String, Object>) e.getOrDefault("metadata", Collections.emptyMap())))
                    .forEach(v -> versions.put(v.getId(), v));
            return versions;
        } else {
            return Collections.emptyMap();
        }
    }

    public static class Builder extends AbstractBuilder<RhoarBooster, RhoarBoosterCatalogService> {
        @Override
        public Builder catalogRef(String catalogRef) {
            return (Builder) super.catalogRef(catalogRef);
        }

        @Override
        public Builder catalogRepository(String catalogRepositoryURI) {
            return (Builder) super.catalogRepository(catalogRepositoryURI);
        }

        @Override
        public Builder pathProvider(BoosterCatalogPathProvider pathProvider) {
            return (Builder) super.pathProvider(pathProvider);
        }

        @Override
        public Builder filter(Predicate<RhoarBooster> filter) {
            return (Builder) super.filter(filter);
        }

        @Override
        public Builder listener(BoosterCatalogListener listener) {
            return (Builder) super.listener(listener);
        }

        @Override
        public Builder transformer(BoosterDataTransformer transformer) {
            return (Builder) super.transformer(transformer);
        }

        @Override
        public Builder environment(String environment) {
            return (Builder) super.environment(environment);
        }

        @Override
        public Builder executor(ExecutorService executor) {
            return (Builder) super.executor(executor);
        }

        @Override
        public Builder rootDir(Path root) {
            return (Builder) super.rootDir(root);
        }

        @Override
        public RhoarBoosterCatalogService build() {
            return new RhoarBoosterCatalogService(this);
        }
    }
}

File: src/main/kotlin/io/fabric8/launcher/booster/catalog/rhoar/RhoarBoosterCatalogService.kt
package io.fabric8.launcher.booster.catalog.rhoar

import io.fabric8.launcher.booster.catalog.AbstractBoosterCatalogService
import io.fabric8.launcher.booster.catalog.Booster
import io.fabric8.launcher.booster.catalog.BoosterFetcher
import io.fabric8.launcher.booster.catalog.spi.BoosterCatalogProvider
import io.fabric8.launcher.booster.catalog.spi.BoosterCatalogSourceProvider
import io.fabric8.launcher.booster.catalog.spi.BoosterMetadataProvider
import java.io.IOException
import java.nio.file.Path
import java.util.*
import java.util.concurrent.ExecutorService
import java.util.function.Predicate
import java.util.function.Supplier
import java.util.logging.Level
import java.util.logging.Logger
import java.util.stream.Collectors
import java.util.stream.Stream
import kotlin.collections.HashMap

class RhoarBoosterCatalogService protected constructor(config: Builder) : AbstractBoosterCatalogService<RhoarBooster>(config), RhoarBoosterCatalog {
    var metadataProvider: BoosterMetadataProvider? = null

    init {
        this.metadataProvider = config.metadataProvider ?: config.discoverMetadataProvider()
    }

    override fun newBooster(data: Map<String, Any?>, boosterFetcher: BoosterFetcher) = RhoarBooster(data, boosterFetcher)

    override fun getMissions() = toMissions(prefilteredBoosters)

    override fun getMissions(filter: Predicate<RhoarBooster>) = toMissions(prefilteredBoosters.filter(filter))

    override fun getRuntimes() = toRuntimes(prefilteredBoosters)

    override fun getRuntimes(filter: Predicate<RhoarBooster>) = toRuntimes(prefilteredBoosters.filter(filter))

    override fun getVersions(filter: Predicate<RhoarBooster>) = toVersions(prefilteredBoosters.filter(filter))

    override fun getVersions(mission: Mission, runtime: Runtime) = getVersions(BoosterPredicates.withMission(mission).and(BoosterPredicates.withRuntime(runtime)))

    override fun getBooster(mission: Mission, runtime: Runtime, version: Version?) =
            prefilteredBoosters
                .filter(BoosterPredicates.withMission(mission))
                .filter(BoosterPredicates.withRuntime(runtime))
                .filter(BoosterPredicates.withVersion(version))
                .findAny()

    private fun toRuntimes(bs: Stream<RhoarBooster>) =
            bs
                .filter { b -> b.runtime != null }
                .map<Runtime>({ it.runtime })
                .collect(Collectors.toCollection(Supplier<TreeSet<Runtime>> { TreeSet() }))

    private fun toMissions(bs: Stream<RhoarBooster>) =
            bs
                .filter { b -> b.mission != null }
                .map<Mission>({ it.mission })
                .collect(Collectors.toCollection(Supplier<TreeSet<Mission>> { TreeSet() }))

    private fun toVersions(bs: Stream<RhoarBooster>) =
            bs
                .filter { b -> b.version != null }
                .map<Version>({ it.version })
                .collect(Collectors.toCollection(Supplier<TreeSet<Version>> { TreeSet() }))

    @Throws(IOException::class)
    override fun indexBoosters(boosters: MutableSet<RhoarBooster>) {
        super.indexBoosters(boosters)

        // Update the boosters with the proper info for missions, runtimes and versions
        val missions = HashMap<String, Mission>()
        val runtimes = HashMap<String, Runtime>()

        // Read the metadata for missions and runtimes
        val metadata = metadataProvider?.invoke()
        if (metadata != null) {
            processMetadata(metadata, missions, runtimes)
        }

        for (booster in boosters) {
            var rname = Objects.toString(booster.metadata["runtime"])
            var r: Runtime? = runtimes[rname]
            if (r == null) {
                r = Runtime(rname ?: "<<missing>>")
                logger.log(Level.WARNING, "Runtime ''{0}'' not found in metadata", r.id)
            }
            booster.runtime = r

            var vname = Objects.toString(booster.metadata["version"])
            var v: Version? = r.versions[vname]
            if (v == null) {
                v = Version(vname ?: "<<missing>>")
                logger.log(Level.WARNING, "Version ''{0}'' not found in Runtime ''{1}'' metadata", arrayOf<Any>(v.id, r.id))
            }
            booster.version = v

            var mname = Objects.toString(booster.metadata["mission"])
            var m: Mission? = missions[mname]
            if (m == null) {
                m = Mission(mname ?: "<<missing>>")
                logger.log(Level.WARNING, "Mission ''{0}'' not found in metadata", m.id)
            }
            booster.mission = m
        }
    }

    /**
     * Process the metadataFile and adds to the specified missions and runtimes
     * maps
     */
    fun processMetadata(metadata: Map<String, Any?>, missions: MutableMap<String, Mission>, runtimes: MutableMap<String, Runtime>) {
        try {
            if (metadata["missions"] is List<*>) {
                val ms = metadata["missions"] as List<Map<String, Any?>>
                ms.stream()
                        .map { e ->
                            Mission(
                                    Objects.toString(e["id"]),
                                    Objects.toString(e["name"]),
                                    Objects.toString(e["description"]),
                                    e.getOrDefault("metadata", emptyMap<Any, Any>()) as Map<String, Any?>)
                        }
                        .forEach { m -> missions[m.id] = m }
            }

            if (metadata["runtimes"] is List<*>) {
                val rs = metadata["runtimes"] as List<Map<String, Any?>>
                rs.stream()
                        .map { e ->
                            Runtime(
                                    Objects.toString(e["id"]),
                                    Objects.toString(e["name"]),
                                    Objects.toString(e["description"]),
                                    e.getOrDefault("metadata", emptyMap<String, Any?>()) as Map<String, Any?>,
                                    Objects.toString(e["icon"]),
                                    getMetadataVersions(e["versions"]))
                        }
                        .forEach { r -> runtimes[r.id] = r }
            }
        } catch (e: Exception) {
            logger.log(Level.SEVERE, "Error while processing metadata", e)
        }

    }

    private fun getMetadataVersions(versionsList: Any?): Map<String, Version> {
        if (versionsList is List<*>) {
            val versions = HashMap<String, Version>()
            val vs = versionsList as List<Map<String, Any?>>
            vs.stream()
                    .map { e ->
                        Version(
                                Objects.toString(e["id"]),
                                Objects.toString(e["name"]),
                                Objects.toString(e["description"]),
                                e.getOrDefault("metadata", emptyMap<Any, Any>()) as Map<String, Any?>)
                    }
                    .forEach { v -> versions[v.id] = v }
            return versions
        } else {
            return emptyMap()
        }
    }

    class Builder : AbstractBoosterCatalogService.AbstractBuilder<RhoarBooster, RhoarBoosterCatalogService>() {
        var metadataProvider: BoosterMetadataProvider? = null

        open fun metadataProvider(metadataProvider: BoosterMetadataProvider): Builder {
            this.metadataProvider = metadataProvider
            return this
        }

        override fun catalogRef(catalogRef: String) = super.catalogRef(catalogRef) as Builder
        override fun catalogRepository(catalogRepositoryURI: String) = super.catalogRepository(catalogRepositoryURI) as Builder
        override fun catalogProvider(catalogProvider: BoosterCatalogProvider) = super.catalogProvider(catalogProvider) as Builder
        override fun sourceProvider(sourceProvider: BoosterCatalogSourceProvider) = super.sourceProvider(sourceProvider) as Builder
        override fun filter(filter: Predicate<RhoarBooster>) = super.filter(filter) as Builder
        override fun listener(listener: (booster: Booster) -> Any) = super.listener(listener) as Builder
        override fun transformer(transformer: (data: Map<String, Any?>) -> Map<String, Any?>) = super.transformer(transformer) as Builder
        override fun executor(executor: ExecutorService) = super.executor(executor) as Builder
        override fun rootDir(root: Path) = super.rootDir(root) as Builder
        override fun build() = RhoarBoosterCatalogService(this)

        internal fun discoverMetadataProvider(): BoosterMetadataProvider =
                discoverCatalogProvider() as BoosterMetadataProvider
    }

    companion object {

        private const val METADATA_FILE = "metadata.json"

        private val logger = Logger.getLogger(RhoarBoosterCatalogService::class.java.name)
    }
}
